src/render/utils/flat-tree.ts

```diff
@@ -0,0 +1,35 @@
+import { addUniqueItem, removeItem } from "../../utils/array"
+import { sortByDepth } from "./sort-by-depth"
+
+export interface DepthChild {
+    depth: number
+}
+
+export class FlatTree {
+    children: DepthChild[] = []
+    numChildren = 0
+
+    private isDirty: boolean = false
+
+    add(child: DepthChild) {
+        addUniqueItem(this.children, child)
+        this.isDirty = true
+    }
+
+    remove(child: DepthChild) {
+        removeItem(this.children, child)
+        this.isDirty = true
+    }
+
+    forEach(callback: (child: DepthChild) => void) {
+        this.isDirty && this.order()
```

Comment:
Not checking the length every `forEach` call looks like a premature optimisation to me. We can remove the `order` function completely by doing:

```suggestion
       const numChildren = this.children.length
       this.isDirty && this.children.sort(compareByDepth)
```